module/zcommon/zfs_fletcher_intel.c

```diff
@@ -0,0 +1,258 @@
+fletcher_4_avx2_init(zio_cksum_t *zcp)
+{
+	ZIO_SET_CHECKSUM(zcp, 0, 0, 0, 0);
+
+	kfpu_begin();
+
+	/* clear avx2 registers */
+	asm volatile("vpxor %ymm0, %ymm0, %ymm0");
+	asm volatile("vpxor %ymm1, %ymm1, %ymm1");
+	asm volatile("vpxor %ymm2, %ymm2, %ymm2");
+	asm volatile("vpxor %ymm3, %ymm3, %ymm3");
+}
+
+static void
+fletcher_4_avx2_fini(zio_cksum_t *zcp)
+{
+	uint64_t __attribute__((aligned(32))) a[4];
+	uint64_t __attribute__((aligned(32))) b[4];
+	uint64_t __attribute__((aligned(32))) c[4];
+	uint64_t __attribute__((aligned(32))) d[4];
+	uint64_t A, B, C, D;
+
+	asm volatile("vmovdqa %%ymm0, %0":"=m" (a));
+	asm volatile("vmovdqa %%ymm1, %0":"=m" (b));
+	asm volatile("vmovdqa %%ymm2, %0":"=m" (c));
+	asm volatile("vmovdqa %%ymm3, %0":"=m" (d));
+
+	kfpu_end();
+
+	A = a[0] + a[1] + a[2] + a[3];
+	B = 0 - a[1] - 2*a[2] - 3*a[3]
+	    + 4*b[0] + 4*b[1] + 4*b[2] + 4*b[3];
+
+	C = a[2] + 3*a[3]
+	    -  6*b[0] - 10*b[1] - 14*b[2] - 18*b[3]
+	    + 16*c[0] + 16*c[1] + 16*c[2] + 16*c[3];
+
+	D = 0 - a[3]
+	    +  4*b[0] + 10*b[1] + 20*b[2] + 34*b[3]
+	    - 48*c[0] - 64*c[1] - 80*c[2] - 96*c[3]
+	    + 64*d[0] + 64*d[1] + 64*d[2] + 64*d[3];
+
+	ZIO_SET_CHECKSUM(zcp, A, B, C, D);
+}
+
+static void
+fletcher_4_avx2(const void *buf, uint64_t size, zio_cksum_t *unused)
+{
+	const uint64_t *ip = buf;
+	const uint64_t *ipend = (uint64_t *)((uint8_t *)ip + size);
+
+	for (; ip < ipend; ip += 2) {
+		asm volatile("vpmovzxdq %0, %%ymm4"::"m" (*ip));
+		asm volatile("vpaddq %ymm4, %ymm0, %ymm0");
+		asm volatile("vpaddq %ymm0, %ymm1, %ymm1");
+		asm volatile("vpaddq %ymm1, %ymm2, %ymm2");
+		asm volatile("vpaddq %ymm2, %ymm3, %ymm3");
+	}
+}
+
+static void
+fletcher_4_avx2_byteswap(const void *buf, uint64_t size, zio_cksum_t *unused)
+{
+	const uint64_t *ip = buf;
+	const uint64_t *ipend = (uint64_t *)((uint8_t *)ip + size);
+	const uint64_t __attribute__((aligned(32))) mask[] = {
+		0xFFFFFFFF00010203, 0xFFFFFFFF08090A0B,
+		0xFFFFFFFF00010203, 0xFFFFFFFF08090A0B };
+
+	asm volatile("vmovdqa %0, %%ymm5"::"m"(*mask));
+
+	for (; ip < ipend; ip += 2) {
+		asm volatile("vpmovzxdq %0, %%ymm4"::"m" (*ip));
+		asm volatile("vpshufb %ymm5, %ymm4, %ymm4");
+
+		asm volatile("vpaddq %ymm4, %ymm0, %ymm0");
+		asm volatile("vpaddq %ymm0, %ymm1, %ymm1");
+		asm volatile("vpaddq %ymm1, %ymm2, %ymm2");
+		asm volatile("vpaddq %ymm2, %ymm3, %ymm3");
+	}
+}
+
+static boolean_t fletcher_4_avx2_valid(void)
+{
+	return (zfs_avx_available() && zfs_avx2_available());
+}
+
+const struct fletcher_4_calls fletcher_4_avx2_calls = {
+	.init = fletcher_4_avx2_init,
+	.fini = fletcher_4_avx2_fini,
+	.compute = fletcher_4_avx2,
+	.compute_byteswap = fletcher_4_avx2_byteswap,
+	.valid = fletcher_4_avx2_valid,
+	.name = "avx2"
+};
+
+#ifdef UNITTEST
+
+/*
+ * Compiled by:
+ *    gcc -O2 -DHAVE_AVX -DHAVE_AVX2 -DUNITTEST \
+ *      -o zfs_fletcher_intel zfs_fletcher_intel.c \
+ *      -I ../../../zfs/include -I ../../../zfs/lib/libspl/include
+ */
+int
+main(int argc, char **argv)
```

Comment:
This is the kind of thing we should always build.  You could make it available as a `fletcher_4_avx2_unit_test()` function have add it to `ztest()` in the `ztest_info` array as a `zopt_rarely` function.  That way it's always getting some run time.